ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/editor/plugin/ExecutionEnvironmentSection.java

```diff
@@ -262,17 +274,53 @@ private void handleRemove() {
 	}
 
 	private void handleAdd() {
-		ElementListSelectionDialog dialog = new ElementListSelectionDialog(PDEPlugin.getActiveWorkbenchShell(), new EELabelProvider());
+		ElementListSelectionDialog dialog = showDialog(true);
+		if (dialog.open() == Window.OK) {
+			addExecutionEnvironments(dialog.getResult());
+		}
+	}
+
+	private void handleUpgrade() {
+		RequiredExecutionEnvironmentHeader header = getHeader();
+		IExecutionEnvironment[] executionEnvironments = JavaRuntime.getExecutionEnvironmentsManager()
+				.getExecutionEnvironments();
+		if (executionEnvironments.length > 0) {
+			ElementListSelectionDialog dialog = showDialog(false);
+			if (dialog.open() == Window.OK) {
+
+				List<String> addedExecutionEnvironments = header.getEnvironments();
+				// Remove existing EE ( Upgrade is allowed only when single Java
+				// EE is present in the header)
+				getHeader().removeExecutionEnvironment(addedExecutionEnvironments.get(0));
+
+				// add selected Java EE to the header
+				addExecutionEnvironments(dialog.getResult());
+
+				final IProject project = getPage().getPDEEditor().getCommonProject();
+				updateClasspathSettings(project);
+			}
+		}
+	}
+
+	private ElementListSelectionDialog showDialog(boolean allowMultiSelection) {
+		ElementListSelectionDialog dialog = new ElementListSelectionDialog(PDEPlugin.getActiveWorkbenchShell(),
+				new EELabelProvider()) {
+			@Override
+			protected FilteredList createFilteredList(Composite parent) {
+				FilteredList filteredList = super.createFilteredList(parent);
+				filteredList.setComparator(VMUtil.ASCENDING_EE_JAVA_VERSION.reversed());
+				return filteredList;
+			}
+		};
 		dialog.setElements(getEnvironments());
 		dialog.setAllowDuplicates(false);
-		dialog.setMultipleSelection(true);
+		dialog.setMultipleSelection(allowMultiSelection);
 		dialog.setTitle(PDEUIMessages.RequiredExecutionEnvironmentSection_dialog_title);
 		dialog.setMessage(PDEUIMessages.RequiredExecutionEnvironmentSection_dialogMessage);
 		dialog.create();
-		PlatformUI.getWorkbench().getHelpSystem().setHelp(dialog.getShell(), IHelpContextIds.EXECUTION_ENVIRONMENT_SELECTION);
-		if (dialog.open() == Window.OK) {
-			addExecutionEnvironments(dialog.getResult());
-		}
+		PlatformUI.getWorkbench().getHelpSystem().setHelp(dialog.getShell(),
+				IHelpContextIds.EXECUTION_ENVIRONMENT_SELECTION);
+		return dialog;
```

Comment:
Handling cancellations and alike here avoids having the logic in all callers.
```suggestion
		if (dialog.open() == Window.OK) {
			return dialog.getResult();
		}
		return null;
```